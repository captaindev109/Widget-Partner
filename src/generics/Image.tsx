import { FC } from 'react';
import styled from 'styled-components';
import { ImageProps } from '@/types/types';

const StyledImage = styled.img<ImageProps>`
  width: ${(props) => props.width};
  height: ${(props) => props.height};
  object-fit: ${(props) => props.fit};
`;

const Image: FC<ImageProps> = ({ src, alt, width, height, fit }) => {
  return <StyledImage src={src} alt={alt} width={width} height={height} fit={fit} />;
};

const defaultProps: Partial<ImageProps> = {
  width: '100%',
  height: '100%',
  fit: 'cover',
};

Image.defaultProps = defaultProps;

export default Image;
