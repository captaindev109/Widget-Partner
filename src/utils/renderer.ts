import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
