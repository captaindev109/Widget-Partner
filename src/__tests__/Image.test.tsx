import { render, RenderResult } from '@testing-library/react';
import queries from '@testing-library/dom/types/queries';
import renderer from 'react-test-renderer';
import Image from '../generics/Image';

describe('Test cases of the Image component', () => {
  let renderedComponent: RenderResult<typeof queries, HTMLElement>;

  const url = 'https://picsum.photos/800/400/';
  const alt = 'test image';
  const child_text = 'child';

  beforeEach(() => {
    renderedComponent = render(
      <Image src={url} alt={alt}>
        <div>{child_text}</div>
      </Image>
    );
  });

  test('to see if the component render without crashing', () => {
    render(<Image src={url} alt={alt} />);
  });

  test('to see if the image src is correct', () => {
    const { getByAltText } = renderedComponent;
    const image = getByAltText(alt);

    expect(image).toHaveAttribute('src', url);
  });

  test('to see if the image alt is correct', () => {
    const { container } = renderedComponent;

    expect(container.querySelector('img')).toHaveAttribute('alt', alt);
  });

  test('to see if the Image can render children', () => {
    const { container } = renderedComponent;

    expect(container.querySelector('div')).toBeNull();
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<Image src={url} alt={alt} />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
