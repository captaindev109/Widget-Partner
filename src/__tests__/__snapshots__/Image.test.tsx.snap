// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cases of the Image component Snapshot test should pass 1`] = `
<img
  alt="test image"
  className="Image__StyledImage-sc-34jhra-0 kDLjWP"
  height="100%"
  src="https://picsum.photos/800/400/"
  width="100%"
/>
`;
